src/neptune_scale/sync/sync_process.py

```diff
@@ -555,3 +579,101 @@ def work(self) -> None:
     @staticmethod
     def _is_fatal_error(ex: Exception) -> bool:
         return isinstance(ex, NeptuneProjectError) or isinstance(ex, NeptuneRunError)
+
+
+class FileUploaderThread(Daemon):
+    def __init__(
+        self, project: str, api_token: str, operations_repository: OperationsRepository, errors_queue: ErrorsQueue
+    ) -> None:
+        super().__init__(name="FileUploaderThread", sleep_time=1)
+
+        self._project = project
+        self._neptune_api_token = api_token
+        self._operations_repository = operations_repository
+        self._errors_queue = errors_queue
+
+        self._neptune_client: Optional[ApiClient] = None
+
+    def close(self) -> None:
+        if self._neptune_client is not None:
+            self._neptune_client.close()
+
+    def work(self) -> None:
+        try:
+            while file_upload_requests := self._operations_repository.get_file_upload_requests(10):
+                logger.debug(f"Have {len(file_upload_requests)} file upload requests to process")
+
+                if not self._neptune_client:
+                    self._neptune_client = backend_factory(self._neptune_api_token)
+
+                target_paths = [file.target_path for file in file_upload_requests]
+
+                try:
+                    storage_urls = fetch_file_storage_urls(self._neptune_client, self._project, target_paths)
+                except NeptuneRetryableError as e:
+                    logger.debug(f"Error while fetching file storage urls: {e}")
+                    self._errors_queue.put(e)
+                    return
+
+                for file in file_upload_requests:
+                    try:
+                        upload_file(file.source_path, file.mime_type, file.size_bytes, storage_urls[file.target_path])
+                    except azure.core.exceptions.AzureError as e:
+                        logger.warning(f"Will retry uploading file {file.source_path}: {e}")
+                        return
+                    except Exception as e:
+                        # Fatal failure. Do not retry the file, but clean it up and continue with the next one
```

Comment:
I agree if the file doesn't exist. What other exceptions should trigger deleting the file from disk / DB? I think we need more granular exception handling here - catching Exception is very broad and the consequence is dropping user's data.